Skip tests if scsi_debug module is already loaded and in use

disgoel opened the pull request:
Several tests across block/, scsi/, dm/, md/, zbd/, nvme/ require exclusive access to the scsi_debug module because they load, unload or reconfigure it. When scsi_debug is already loaded by the environment (e.g., by another driver or a previous setup), these tests fail with:
```
modprobe: FATAL: Module scsi_debug is in use.
Unloading scsi_debug failed
```

```
# lsmod | grep scsi_debug
scsi_debug            327680  4
```
Instead of modifying common rc files—which would overskip unrelated tests, this patch adds `_module_not_in_use scsi_debug` only to the tests that actually depend on exclusive access to scsi_debug.

disgoel commented:
> I took a look in the commit, but it looks missing the implementation of _module_not_in_use() function, doesn't it?

The function already exists.
https://github.com/linux-blktests/blktests/blob/master/common/rc#L95

disgoel commented:
> @disgoel Thanks for the idea. It is interesting for me, and I would like to think about it. When the scsi_debug is a loadable module, it makes sense that the scsi_debug is not loaded before running each of the test cases, probably. I wonder how the check can be done when the scsi_debug is built-in.

For built-in scsi_debug, the situation is actually more constrained than the modular case. Since the driver cannot be unloaded or reset, any existing scsi_debug devices mean the tests cannot obtain exclusive control. In that case, skipping the tests is still the correct and safest behavior.

The intent of the check is not to distinguish built-in vs modular scsi_debug, but to ensure exclusive access. If that exclusivity cannot be guaranteed (either because the module is already in use or because the driver is built-in and active), the test should be skipped.

kawasaki reviewed (changes requested):
@disgoel Thank you for updating the patch. I agree with this fix concept. Please find my review comments and see if the comments make sense for you.

For your reference, I share a code change which reflects my review comments ([link](https://github.com/kawasaki/blktests/commit/463b9895693f5e422d35fcb96d877b583bf97b2c)). I hope it helps to clarify my intents.

Also, I think some more test cases can be modified to use _have_loadable_scsi_debug or _have_scsi_debug. I created another code change to try it out ([link](https://github.com/kawasaki/blktests/commit/4c30cbd694149908f9ec1cca14d6879d17ac8aeb)). Please take a look in it also.

kawasaki commented:
@disgoel Thanks again for your effort. I think this work is close to the merge. Before the merge, I would like to post your patch to the linux-block mailing list to get wider reviews and build up consensus. I can post the patch to the list on behalf of you with your authorship. Just in case this is not okay for you, please let me know.

disgoel commented on common/scsi_debug:
```diff
@@ -5,7 +5,25 @@
-	_have_driver scsi_debug
+	_have_module scsi_debug
+	if [[ -d /sys/module/scsi_debug ]]; then
+		_module_not_in_use scsi_debug || return 1
```
made changes accordingly

disgoel commented:
@kawasaki Thank you for the detailed review and the reference code. I have updated the PR to address requested changes. Also updated the suggested test cases (block/009, md/002, etc.) to use the appropriate helper.

I am completely fine with you posting the patch to the linux-block mailing list on my behalf with my authorship. I appreciate the help in getting this merged!

kawasaki commented:
@disgoel Thank you for updating the patch. While I did the final check of the patch, I noticed that it does not work as expected. Agrrr.

1. I overlooked the behavior of /sysfs/module/scsi_debug/refcnt is not working as expected. It is just zero even afater a scsi_debug host is created. I should have noticed that at the beginning. My bad...
2. It does not work for the test case which creates scsi_debug device as "fallback". e.g. scsi/009.
3. It does not work well when other helper function loaded scsi_debug. e.g. "_have_module_param scsi_debug inq_vector" call in scsi/005.
4. srp group test cases, zbd/012 and _have_scsi_debug_group_number_stats() also use loadable scsi_debug. They should check _have_loadable_scsi_debug.

I created some patches to address these problems, which are available in the branch [here](https://github.com/kawasaki/blktests/commits/fix-scsi-debug/). I will do some more testing and will post these as a series to linux-block.

Your patch is [here](https://github.com/kawasaki/blktests/commit/3dcc3251b33fe43ad380a75638f5d77e96abeded), in the series. I'm thinking to fold-in following two changes [1][2].

[1] https://github.com/kawasaki/blktests/commit/11656fdc5893d96c56016ba383be8a350642ad02
[2] https://github.com/kawasaki/blktests/commit/ec0853fe70d95fe6d7d9cd535dadfa27fb7d5864

Especially, [1] is the key difference. Could you check these changes?
